### DISCUSSION TITLE
Callback scheduling

### DISCUSSION THREAD

**Discussion by IFX-Anusha** [micropython/micropython#13107] [Core Development] [answered]
Hi,

We are working on the psoc6 port. In that, the GPIO interrupt function is presumably blocked by the UART.  Because after running the code from thonny/repl ,  the callback function is only executed after giving an input character from the repl. To solve this I also have added   MICROPY_EVENT_POLL_HOOK in the  mp_hal_stdin_rx_chr(void) function as below, 

```
int mp_hal_stdin_rx_chr(void) {
    for(;;){
    uint8_t c = 0;
    cy_rslt_t result;
    result = cyhal_uart_getc(&cy_retarget_io_uart_obj, &c, 0);
    if(result == CY_RSLT_SUCCESS){
    return c;
    }
    MICROPY_EVENT_POLL_HOOK
    }
}
```
Could anyone help me by giving more insights on the  mp_sched_schedule() function with any possible solutions? Thanks in Advance!!

**Comment by sophiedegran** [comment]
Have you look at github.com/peterhinch/micropython-async/tree/master/v3/as_demo/auart.py ? Don't know if that help you... but I think async mode can be a good solution

**Reply by IFX-Anusha** [reply]
Thanks for the reply!!. But this is not what I am looking for. I would like to understand about the mp_sched_schedule() function and the issue I mentioned above.

**Comment by andrewleech** [comment, marked as answer]
The scheduler is handled in `MICROPY_EVENT_POLL_HOOK` so yes it's common practice to call that in any idle loop like waiting for input and sleep functions.

See:
* https://github.com/micropython/micropython/blob/30b0ee34d3a7cf6f967387e783268c8a939e04c8/ports/stm32/mphalport.c#L38
* https://github.com/micropython/micropython/blob/30b0ee34d3a7cf6f967387e783268c8a939e04c8/ports/esp32/mphalport.c#L115

**Reply by IFX-Anusha** [reply]
Thankyou! Now it's clear.
